File: AMENAPP/Backend/functions/src/contracts/contextSignal.ts
export type SignalType =
    | "noteSaved"
    | "noteThemeDetected"
    | "churchSaved"
    | "churchUnsaved"
    | "churchViewed"
    | "visitVerified"
    | "prayerCreated"
    | "prayerAnswered"
    | "prayerReminderActed"
    | "giftCompleted"
    | "givingCauseViewed"
    | "messageSent"
    | "prayerExtractedFromMessage"
    | "studyStarted"
    | "studyCompleted"
    | "verseReflected"
    | "wellnessToolUsed"
    | "crisisSurfaceOpened"
    | "groupJoined"
    | "eventRSVPed"
    | "volunteerMatched"
    | "sessionRhythmTick";

export type TierCeiling = "s" | "c" | "p";

export type GraphNodeType =
    | "church"
    | "churchVisit"
    | "sermon"
    | "note"
    | "prayerRequest"
    | "person"
    | "group"
    | "community"
    | "gift"
    | "cause"
    | "event"
    | "study"
    | "verse"
    | "skill"
    | "interest"
    | "goal"
    | "milestone"
    | "wellnessActivity"
    | "reflection";

export type ConsentEdge =
    | "notesToMatching"
    | "notesToGiving"
    | "messagesToPrayer"
    | "locationToVisits"
    | "givingToFeed"
    | "activityToCheckIns"
    | "graphToBerean"
    | "graphToCohorts"
    | "activityToRhythm"
    | "crossDeviceContinuity";

export interface GraphRef {
    nodeType: GraphNodeType;
    nodeID: string;
}

export type AnyCodableValue =
    | string
    | number
    | boolean
    | null
    | AnyCodableValue[]
    | { [key: string]: AnyCodableValue };

export interface ContextSignal {
    id: string;
    type: SignalType;
    tierCeiling: TierCeiling;
    subjectRefs: GraphRef[];
    payload: { [key: string]: AnyCodableValue };
    occurredAt: string;
    decayHalfLifeDays: number;
    consentEdgeRequired?: ConsentEdge | null;
}

export const signalTypes: SignalType[] = [
    "noteSaved",
    "noteThemeDetected",
    "churchSaved",
    "churchUnsaved",
    "churchViewed",
    "visitVerified",
    "prayerCreated",
    "prayerAnswered",
    "prayerReminderActed",
    "giftCompleted",
    "givingCauseViewed",
    "messageSent",
    "prayerExtractedFromMessage",
    "studyStarted",
    "studyCompleted",
    "verseReflected",
    "wellnessToolUsed",
    "crisisSurfaceOpened",
    "groupJoined",
    "eventRSVPed",
    "volunteerMatched",
    "sessionRhythmTick"
];

export function assertServerEmittable(signal: ContextSignal): void {
    if (signal.tierCeiling === "s") {
        throw new Error("Tier-S context signals must never be ingested by server functions.");
    }
}

File: AMENAPP/Shared/Contracts/ContextSignal.swift
import Foundation

enum SignalType: String, Codable, CaseIterable, Sendable {
    case noteSaved
    case noteThemeDetected
    case churchSaved
    case churchUnsaved
    case churchViewed
    case visitVerified
    case prayerCreated
    case prayerAnswered
    case prayerReminderActed
    case giftCompleted
    case givingCauseViewed
    case messageSent
    case prayerExtractedFromMessage
    case studyStarted
    case studyCompleted
    case verseReflected
    case wellnessToolUsed
    case crisisSurfaceOpened
    case groupJoined
    case eventRSVPed
    case volunteerMatched
    case sessionRhythmTick
}

enum TierCeiling: String, Codable, CaseIterable, Sendable {
    case s
    case c
    case p
}

struct ContextSignal: Codable, Identifiable, Equatable, Sendable {
    let id: UUID
    let type: SignalType
    let tierCeiling: TierCeiling
    let subjectRefs: [GraphRef]
    let payload: [String: AnyCodableValue]
    let occurredAt: Date
    let decayHalfLifeDays: Double
    let consentEdgeRequired: ConsentEdge?
}

struct GraphRef: Codable, Hashable, Sendable {
    let nodeType: GraphNodeType
    let nodeID: String
}

enum GraphNodeType: String, Codable, CaseIterable, Sendable {
    case church
    case churchVisit
    case sermon
    case note
    case prayerRequest
    case person
    case group
    case community
    case gift
    case cause
    case event
    case study
    case verse
    case skill
    case interest
    case goal
    case milestone
    case wellnessActivity
    case reflection
}
